### ISSUE TITLE
support for other languages via SWIG

### ISSUE THREAD

**Issue by rkuhn** [automerge/automerge-rs#411]
_Background: I’ve been following @ept’s work for a while now and finally have a defensible reason for joining the effort: I’d like to offer shared documents with automerge semantics to a C# audience, with Actyx as the dissemination layer. This issue may turn out actionable in itself or more of a discussion, not sure yet._

automerge-c looks like the right way for integration of automerge into all kinds of other languages, including strongly typed ones like C#. Writing the glue code is error-prone and a hassle, though, so using a generator like SWIG seems hardly avoidable. If we agree on this, we should make some rules for how the C API is structured so that it is efficient and effective to use correctly — I think there are some improvements we can make over the current state, which to my understanding is still young and not yet fully fledged.

For reference, this is the SWIG interface definition I composed over the past few days while trying out C# support:
```
%module automerge
%{
    #include "automerge.h"
%}

%include "typemaps.i"
%include "stdint.i"
%include "carrays.i"

/* getting raw bytes out of a change */
%array_functions(unsigned char, bytes);

%define ptrdiff_t
intptr_t
%enddef

%include "automerge.h"

%inline %{
const AMobjId *const ROOT = AM_ROOT;
const AMchangeHashes *const NOW = NULL;
%}
```

Some things that are not supported or rather awkward with the current interface are:

- AMresult holds the memory for AMvalue, but obtaining the latter requires separate allocations (I’d propose to fold one into the other)
- getting the raw bytes out of a AMchange is very inefficient
- getting raw bytes into an AMchanges is impossible

As a remark: the Rust implementation of automerge-c seems to use quite a bit more of raw pointers, slice::from_raw_parts, etc. than necessary, but before proposing changes I’d like to understand the full picture.

---

To my mind, it would be very beneficial, both for the implementation as well as for API consumers, to have a rather small but rigorous set of rules governing who owns data and how they are passed between Rust and C (once this issue yields a result, I volunteer for recording it in a design document within the repository). To get things started, here’s my proposal:

- primitive non-pointer C types (i.e. all non-structs) are passed in their native form, by value
- everything else is passed by reference, i.e. strongly typed C pointers (like `const AMresult *`)
- all heap allocations behind those pointers are done by Rust code, implying that functions are provided to create all needed API inputs, and also that the structures are opaque for C code (like e.g. `struct AMactorId` in the currently generated `automerge.h`)

Some handles — like AMdoc — could support cloning, meaning that their backing storage is an `Arc`. The C API users won’t need to know or care, they just need to free all handles when they’re done with them, regardless of their source, but with type-safe disposal functions as today.

Passing e.g. changes in encoded form between C# and automerge would benefit highly from pairs of functions that used byte arrays as in/out parameters so that the CLR’s native marshalling infrastructure (or pinning) can be used.

---

Once a set of rules like the above is established, we can make better decisions on how to represent different items (like object IDs, actor IDs, etc.) — they could become primitive non-pointer C types in the context of a document, for example.

What do you guys think? I’d be interested in a video chat to get to know you and discuss these things as well.

**Comment by ept** [comment]
@jkankiewicz would be great to get your thoughts on this

**Comment by jkankiewicz** [comment]
@rkuhn, thank you for your feedback on the C API! 

> automerge-c looks like the right way for integration of automerge into all kinds of other languages, including strongly typed ones like C#.

Yes, it's intended to provide a base upon which bindings for multiple languages can be built.

> Writing the glue code is error-prone and a hassle, though, so using a generator like SWIG seems hardly avoidable.

Based upon my past work experience, I agree that using a wrapper generator to produce language bindings is preferable to writing them by hand.
SWIG may cover the most programming languages of them all but its generality usually requires a user to do more work than a wrapper generator specific to the targeted language would.
For example, a SWIG-generated Python wrapper for a C++ class must itself be wrapped by a hand-written pure-Python class if it's to be inherited from.

> AMresult holds the memory for AMvalue, but obtaining the latter requires separate allocations...

The C API only copies primitive values wholesale from Rust to C whereas composite values are referenced through pointers so that no copying is necessary.
An `AMbyteSpan` value provides access to an array of primitive values and an iterator value (e.g. `AMchanges`, `AMchangeHashes`, etc.) provides access to a vector of composite values.
In addition to its discriminator enum tag's byte, an `AMvalue` union requires the allocation of four pointers worth of memory in order to accommodate an original iterator value whose state is stored by the C runtime.
The state of an iterator value is stored by the C runtime so that additional iterator values, both forward and reverse, referencing the same Rust vector may be created from an original as necessary in order to access its elements in either direction without requiring additional allocations by the Rust runtime.
 
> AMresult holds the memory for AMvalue, but...I’d propose to fold one into the other...

The C API's functions don't take `AMresult` arguments in order to leverage static type checking by the C compiler and to allow them to directly return primitive values such as the length of a vector.
If a C API function were to take an `AMresult` argument then it would have to check if it contains an error message or an unexpected type of value and it would also have to wrap its result value in an `AMresult` in order to enable it to report that the argument was an error message or a value of an unexpected type.
The bifurcation of `AMresult` and `AMvalue` encourages the detection of an erroneous `AMresult` to take place only once and within the context where it can actually be handled.

> getting the raw bytes out of a AMchange is very inefficient

The `AMchangeRawBytes()` function returns an `AMbyteSpan` which contains a pointer to the first byte in the array of bytes and the array's length.
This approach enables the byte array to be inspected without copying it first.
An `AMbyteSpan`'s fields, `src` and `count`, are named after their corresponding arguments of the standard C `memcpy` function and can be passed directly into the `memchr`, `memcmp`, `memcpy` and `memmove` functions.
If the `AMchangeRawBytes()` function were to instead take `src` as an out parameter and `count` as an in/out parameter then it would have to be called twice in order to obtain the size of the array that must be allocated to receive those bytes e.g.
```
size_t count;
AMchangeRawBytes(change, NULL, &count);
uint8_t* dest = malloc(count);
AMchangeRawBytes(change, dest, &count);
```

@rkuhn, can you please explain in what way(s) it's inefficient?

> getting raw bytes into an AMchanges is impossible

I can add an `AMchangesInit()` function like the `AMchangeHashesInit()` function for creating a `Vec<automerge::Change>` from within C.

> As a remark: the Rust implementation of automerge-c seems to use quite a bit more of raw pointers, slice::from_raw_parts, etc. than necessary...

@rkuhn, can you please provide more detail regarding why that is so?

**Comment by rkuhn** [comment]
Thanks for the detailed response!

Regarding AMresult/AMvalue: when viewed from C as the target language the current approach makes sense, but at least SWIG does not support passing structs by value, so obtaining an AMvalue always incurs an extra allocation (and requires the matching `free()`, which is handled by the generated SWIG code in C#).

I think instead of returning AMresult in many places we can be a lot more specific: AMchanges will only ever return an error or changes, never a document. One way to model this in C is to return an integer discriminant and place the actual value/pointer at a location provided by the caller.

AMbyteSpan may look efficient to C, but it is very inefficient to use from C# as far as I was so far able to figure out, hence my proposal for adding an API where the caller provides a pointer to which the bytes will be copied (i.e. an `out` function argument) — this is supported by C# native interop facilities.

One place with more unsafety than seems required are the `details` fields — I’m not even sure that the current treatment is “defined behaviour” for Rust (in contrast to “it happens to work for now”; just using the same bytes for the same things is not enough for Rust compiler internals, e.g. references need to have the right provenance etc.; have you tried running this with MIRI?).

**Comment by jkankiewicz** [comment]
> ...SWIG does not support passing structs by value...

@rkuhn, could it be that SWIG isn't the best generator to use in this case?

I didn't have SWIG in mind whilst designing the C API because, according to everything I've read, it's not really suitable for generating OO wrapper classes from a C library.
SWIG is suitable for generating OO wrapper classes from a C++ library because its classes and their methods have already been defined but with a C library one has to supply hand-written class and method definitions within the SWIG interface file and, as you wrote, "writing the glue code is error-prone and a hassle".

Have you considered using [interoptopus](https://docs.rs/interoptopus/latest/interoptopus/) to generate C# bindings for automerge-rs instead?

While it's true that interoptopus will still require you to supply hand-written C# class and method definitions, at least the OO wrapper classes that it generates will be based directly upon the underlying Rust library's structs, methods and functions and should thereby skirt the inefficiencies imposed by going through an intermediate C library.

**Comment by alexjg** [comment]
My two cents: I think it would be preferable to base any language bindings on the Rust code directly rather than on the C bindings. My reasoning for this is that this project is primarily a Rust project. This means that:

* The Rust codebase is the actual implementation. Any changes made to the library will be made there. Any language binding which depends on the C codebase would need to wait for @jkankiewicz to update the C bindings.
* Relatedly, most of the contributors on this project are Rust developers. Maintaining a language binding written against the Rust API just requires us to have one Rust maintainer and one $LANGUAGE maintainer whilst depending on the C binding means we have an additional dependency. In practice this means that @jkankiewicz could end up becoming a bottleneck for any changes to dependent bindings.
* It's much easier to debug language bindings when they talk directly to the Rust code. Any additional layers make debugging much harder. 

For these reasons I would be quite hesitant about accepting contributions to this repository which add language bindings based on `automerge-c`. That said, I am very eager to see additional language bindings based on the Rust codebase - easy interop is one of the reasons I'm excited about automerge-rs. At first blush Interoptus looks like it would do the trick, @rkuhn what do you think of it?

**Comment by jkankiewicz** [comment]
> [The C API is] intended to provide a base upon which bindings for multiple languages can be built.

@rkuhn, I should have instead written that the C API is intended to provide a base upon which bindings for a language can be built *when that language lacks a more direct avenue for doing so*.

>  [W]hen viewed from C as the target language the current approach makes sense...

The C API has been designed for use by C programmers if for no other reason than I have to port a lot of unit tests from Rust and WASM to C in order to validate its correctness. The `AMresultStack` struct and its related functions, for example, ease memory management but they shouldn't be exposed to an OO language where it makes more sense to wrap an `AMresult*` in a smart pointer object that calls `AMfree()` on it when it goes out of scope.

**Comment by rkuhn** [comment]
Thanks for the clarification: automerge-c is _not_ “intended to be an ffi integration point for all other languages” as stated by the main README, so I think that should be changed. I still think that going via a C interface makes a lot of sense because every other language out there has C interop facilities while only a very select few provide Rust interop. Not sure what I’ll do, then, but it probably won’t end up in this repo.
